aws/resource_aws_kinesis_analytics_application_test.go

```diff
@@ -0,0 +1,426 @@
+		resp, err := conn.DescribeApplication(describeOpts)
+		if err != nil {
+			return err
+		}
+
+		*application = *resp.ApplicationDetail
+
+		return nil
+	}
+}
+
+func testAccKinesisAnalyticsApplication_basic(rInt int) string {
+	return fmt.Sprintf(`
+resource "aws_kinesis_analytics_application" "test" {
+  name = "testAcc-%d"
+  code = "testCode\n"
+}
+`, rInt)
+}
+
+func testAccKinesisAnalyticsApplication_update(rInt int) string {
+	return fmt.Sprintf(`
+resource "aws_kinesis_analytics_application" "test" {
+  name = "testAcc-%d"
+  code = "testCode2\n"
+}
+`, rInt)
+}
+
+func testAccKinesisAnalyticsApplication_cloudwatchLoggingOptions(rInt int, streamName string) string {
+	return fmt.Sprintf(`
+resource "aws_cloudwatch_log_group" "test" {
+  name = "testAcc-%d"
+}
+
+resource "aws_cloudwatch_log_stream" "test" {
+  name = "testAcc-%s"
+  log_group_name = "${aws_cloudwatch_log_group.test.name}"
+}
+
+resource "aws_kinesis_analytics_application" "test" {
+  name = "testAcc-%d"
+  code = "testCode\n"
+
+  cloudwatch_logging_options {
+    log_stream = "${aws_cloudwatch_log_stream.test.arn}"
+    role = "${aws_iam_role.test.arn}"
+  }
+}
+`, rInt, streamName, rInt)
+}
+
+func testAccKinesisAnalyticsApplication_inputsKinesisStream(rInt int) string {
+	return fmt.Sprintf(`
+resource "aws_kinesis_stream" "test" {
+  name = "testAcc-%d"
+  shard_count = 1
+}
+
+resource "aws_kinesis_analytics_application" "test" {
+  name = "testAcc-%d"
+  code = "testCode\n"
+
+  inputs {
+    name_prefix = "test_prefix"
+    kinesis_stream {
+      resource = "${aws_kinesis_stream.test.arn}"
+      role = "${aws_iam_role.test.arn}"
+    }
+    parallelism {
+      count = 1
+    }
+    schema {
+      record_columns {
+        mapping = "$.test"
+        name = "test"
+        sql_type = "VARCHAR(8)"
+      }
+      record_encoding = "UTF-8"
+      record_format {
+        mapping_parameters {
+          json {
+            record_row_path = "$"
+          }
+        }
+      }
+    }
+  }
+}
+`, rInt, rInt)
+}
+
+func testAccKinesisAnalyticsApplication_inputUpdateKinesisStream(rInt int, streamName string) string {
+	return fmt.Sprintf(`
+resource "aws_kinesis_stream" "test" {
+  name = "testAcc-%s"
+  shard_count = 1
+}
+
+resource "aws_kinesis_analytics_application" "test" {
+  name = "testAcc-%d"
+  code = "testCode\n"
+
+  inputs {
+    name_prefix = "test_prefix2"
+    kinesis_stream {
+      resource = "${aws_kinesis_stream.test.arn}"
+      role = "${aws_iam_role.test.arn}"
+    }
+    parallelism {
+      count = 2
+    }
+    schema {
+      record_columns {
+        mapping = "$.test2"
+        name = "test2"
+        sql_type = "VARCHAR(8)"
+      }
+      record_encoding = "UTF-8"
+      record_format {
+        mapping_parameters {
+          csv {
+            record_column_delimiter = ","
+            record_row_delimiter = "\n"
+          }
+        }
+      }
+    }
+  }
+}
+`, streamName, rInt)
+}
+
+func fulfillSleep() resource.TestCheckFunc {
```

Comment:
I put in a sleep here because I found that the trust policy for the IAM role needs to propagate before referencing the role in the tests

Reply:
Hi @kl4w 👋  We prefer that the resource usually handles this type of logic by retrying on the specific error message(s) that are returned for a minute or two. e.g.

```go
// Handle IAM eventual consistency
err := resource.Retry(1*time.Minute, func() *resource.RetryError {
  // CreateApplication as an example
  err := conn.CreateApplication(input)
  if err != nil {
    // kinesisanalytics.ErrCodeInvalidArgumentException as an example
    if isAWSErr(err, kinesisanalytics.ErrCodeInvalidArgumentException, "some IAM-related error message, e.g. not authorized") {
      return resource.RetryableError(err)
    }
    return resource.NonRetryableError(err)
  }
  return nil
})
if err != nil {
  return fmt.Errorf("error creating application: %s", err)
}
```